README.md
# node-red-contrib-tjfs-nodes
[![Platform](https://img.shields.io/badge/platform-Node--RED-red)](https://nodered.org)
<br>[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)
[![GitHub license](https://img.shields.io/github/license/dceejay/tfjs-nodes)](https://github.com/dceejay/tfjs-nodes/blob/master/LICENSE)

A `Node-RED` node for `tensorflow.js`. A simple function like node preloaded with tensorflowjs and some other useful libraries.

## Install
You have two options to install the node.
 * Use `Manage palette` option in `Node-RED` Menu
 * Run the following command in your Node-RED user directory - typically `~/.node-red`
 ```
 npm install node-contrib-tfjs-nodes
 ```
**Note:** You need to restart `Node-RED` after installation. If installation goes wrong please open a [new issue](https://github.com/dceejay/tfjs-nodes/issues/new/choose).

## Overview
This package contains 4 nodes
 - **tf predict** - Runs tf.predict against any tensorflow model you import
 - **tf mobilenet** - Runs the pretrained mobilenet classifier
 - **tf coco ssd** - Runs the COCO Single Shot Detector against a file or a buffer of a jpg image
 - **tf posenet** - Runs the Posenet body parts pretrained model against a file or a buffer of a jpg image

**Note:** `tensorflow.js` is only available in 64-bit, so it will not run on 32-bit operating systems like Raspbian.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]
### Added
 - New `tf mobilenets` node
 - Standarized node status defined globally for all nodes
 - `JavaScript` code has been standarized following [Standard JS](https://standardjs.com/index.html)
 - Badges in `README` (platform Node-RED, code style, licence)
 - `CHANGELOG` file
 - New contributor `Yair Bonastre` in `package.json`
 - `google` keyword to `package.json` 
 - New `Parameters` section in configuration
 - Added `Online` and `Local` options for all nodes (`tf coco ssd` and `tf mobilenets` only supported)
 - `.eslintrc.js` file with `Standard JS` reference
 - Override of messages properties (threshold and max. detections)
 - Two functions to filter and count classes
 - Call of `tf.dispose` to solve possible memory leaks after every inference

### Changed
 - General polish `JavaScript` code across all nodes
 	- Variable names
 	- Refining general reusable functions
 - Tensorflow node logo updated to 2.0
 - Node color matches official Tensorflow orange color
 - Node-RED example (improved)
 - Better icons in nodes configuration
 - Standard configuration templates
 - Revised information help menus 
 - Name of npm module to `node-red-contrib-tfjs-nodes` and its description
 - JavaScript file from `tfjs.js` to `tfjs-nodes.js`
 - Set all used libraries to `const` variables
 - `Min. Threshold` and `Max. Detections` are now selected from an easier spinner
 - `Min. Threshold` is now defined from 0 to 100 in %
 - Refactor of the variable types whether to use `var`, `let` or `const` 
 - `@tensorflow-models/coco-ssd` version upgraded to `v2.0.3`

### Fixed
 - Correct model initial loading for `tf coco ssd` node 


The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

tfjs.js

module.exports = function (RED) {
    function TensorFlowPredict(n) {
        var fs = require('fs');
        /* suggestion from https://github.com/tensorflow/tfjs/issues/2029 */
        const nodeFetch = require('node-fetch'); // <<--- ADD
        global.fetch = nodeFetch; // <<--- ADD
        /* ************************************************************** */
        var tf = require('@tensorflow/tfjs-node');
        
        RED.nodes.createNode(this, n);
        this.modelfile = n.model;
        //this.modelfile = 'https://storage.googleapis.com/tfjs-models/tfjs/mobilenet_v1_0.25_224/model.json';
        var node = this;

        async function loadModel(m) {
            await tf.loadLayersModel(m).then(function (model) {
                var shp = model.inputs[0].shape;
                shp.shift();
                node.log("Shape wanted: "+shp);
                shp.unshift(1);
                model.predict(tf.zeros(shp)).dispose();
                node.model = model;
                node.shp = shp;
                node.ready = true;
                node.status({fill:'green', shape:'dot', text:'Model ready'});
            });
        }
        node.status({fill:'yellow', shape:'ring', text:'Loading model...'});
        loadModel(node.modelfile);

        async function imgToTensor(p) {
            // if it's a string assume it's a filename
            if (typeof p === "string") { p = fs.readFileSync(p); }
            const img = tf.node.decodeImage(p,node.shp[3]);
            // rescale the image to fit the wanted shape
            const scaled = tf.image.resizeBilinear(img, [node.shp[1],node.shp[2]], true);
            const offset = tf.scalar(127.5);
            // Normalize the image from [0, 255] to [-1, 1].
            const normalized = scaled.sub(offset).div(offset);
            // extend the tensor to 4d
            return normalized.reshape(node.shp);
        }

        node.on('input', function (msg) {
            msg.inputShape = node.shp;
            try {
                imgToTensor(msg.payload)
                    .then(
                        function(t) {
                            var tensorResult = node.model.predict(t);
                            msg.maxIndex = tensorResult.argMax(1).dataSync()[0];
                            msg.payload = Array.from(tensorResult.dataSync());
                            node.send(msg);
                        },
                        function(e) {
                            node.error("OOPS: "+e,msg);
                        }
                    )
            } catch (error) {
                node.error(error, msg);
            }
        });

        node.on("close", function () {
            node.status({});
        });
    }
    RED.nodes.registerType("tensorflowPredict", TensorFlowPredict);

    function TensorFlowCoCo(n) {
        var fs = require('fs');
        var express = require("express");
        var compression = require("compression");

        /* suggestion from https://github.com/tensorflow/tfjs/issues/2029 */
        const nodeFetch = require('node-fetch'); // <<--- ADD
        global.fetch = nodeFetch; // <<--- ADD
        /* ************************************************************** */

        var tf = require('@tensorflow/tfjs-node');
        var cocoSsd = require('@tensorflow-models/coco-ssd');
        
        RED.nodes.createNode(this, n);
        this.scoreThreshold = n.scoreThreshould;
        this.maxDetections = n.maxDetections;
        this.passthru = n.passthru || false;
        this.modelUrl = n.modelUrl || undefined; // "http://localhost:1880/coco/model.json"
        var node = this;

        RED.httpNode.use(compression());
        RED.httpNode.use('/coco', express.static(__dirname + '/models/coco-ssd'));

        async function loadModel() {
            node.model = await cocoSsd.load({modelUrl: node.modelUrl});
            node.ready = true;
            node.status({fill:'green', shape:'dot', text:'Model ready'});
        }
        node.status({fill:'yellow', shape:'ring', text:'Loading model...'});
        loadModel();

        node.on('input', function (msg) {
            async function reco(img) {
                msg.maxDetections = msg.maxDetections || node.maxDetections || 20;
                msg.payload = await node.model.detect(img, msg.maxDetections);
                msg.shape = img.shape;
                msg.classes = {};
                msg.scoreThreshold = msg.scoreThreshold || node.scoreThreshold || 0.5;
                
                for (var i=0; i<msg.payload.length; i++) {
                    if (msg.payload[i].score < msg.scoreThreshold) {
                        msg.payload.splice(i,1);
                        i = i - 1;
                    }
                }
                for (var j=0; j<msg.payload.length; j++) {
                    msg.classes[msg.payload[j].class] = (msg.classes[msg.payload[j].class] || 0 ) + 1;
                }
                node.send(msg);
            }
            try {
                if (node.ready) {
                    var p = msg.payload;
                    if (typeof p === "string") { p = fs.readFileSync(p); }
                    if (node.passthru === true) { msg.image = p; }
                    reco(tf.node.decodeImage(p));
                }
            } catch (error) {
                node.error(error, msg);
            }
        });

        node.on("close", function () {
            node.status({});
        });
    }
    RED.nodes.registerType("tensorflowCoco", TensorFlowCoCo);
    
   
    function TensorFlowPose(n) {
        var fs = require('fs');
        /* suggestion from https://github.com/tensorflow/tfjs/issues/2029 */
        const nodeFetch = require('node-fetch'); // <<--- ADD
        global.fetch = nodeFetch; // <<--- ADD
        /* ************************************************************** */
        var tf = require('@tensorflow/tfjs-node');
        var posenet = require('@tensorflow-models/posenet');
        //import * as posenet from '@tensorflow-models/posenet';
        
        RED.nodes.createNode(this, n);
        this.scoreThreshold = n.scoreThreshould;
        this.maxDetections = n.maxDetections;
        var node = this;

        async function loadModel() {
            node.model = await posenet.load();
            node.ready = true;
            node.status({fill:'green', shape:'dot', text:'Model ready'});
        }
        node.status({fill:'yellow', shape:'ring', text:'Loading model...'});
        loadModel();

        node.on('input', function (msg) {
            async function estimateMultiplePosesOnImage(image) {
                msg.scoreThreshold = msg.scoreThreshold || node.scoreThreshold || 0.5;
                msg.maxDetections = msg.maxDetections || node.maxDetections || 4;
                const poses = await node.model.estimateMultiplePoses(image, {
                    flipHorizontal: false,
                    maxDetections: msg.maxDetections,
                    scoreThreshold: msg.scoreThreshold,
                    nmsRadius: 20
                });
                msg.payload = poses;
                for (var i=0; i<msg.payload.length; i++) {
                    if (msg.payload[i].score < msg.scoreThreshold) {
                        msg.payload.splice(i,1);
                        i = i - 1;
                    }
                }
                msg.shape = image.shape;
                msg.classes = {person:msg.payload.length};
                node.send(msg);
            }
            try {
                if (node.ready) {
                    var p = msg.payload;
                    if (typeof p === "string") { p = fs.readFileSync(p); }
                    estimateMultiplePosesOnImage(tf.node.decodeImage(p));
                }
            } catch (error) {
                node.error(error, msg);
            }
        });

        node.on("close", function () {
            node.status({});
        });
    }
    RED.nodes.registerType("tensorflowPose", TensorFlowPose);
};


getModels.sh
#!/bin/bash

echo "Get CoCo SSD model"
rm -f models/coco-ssd/*
wget -q --show-progress -P models/coco-ssd/ https://storage.googleapis.com/tfjs-models/savedmodel/ssdlite_mobilenet_v2/model.json
wget -q --show-progress -P models/coco-ssd/ https://storage.googleapis.com/tfjs-models/savedmodel/ssdlite_mobilenet_v2/group1-shard1of5
wget -q --show-progress -P models/coco-ssd/ https://storage.googleapis.com/tfjs-models/savedmodel/ssdlite_mobilenet_v2/group1-shard2of5
wget -q --show-progress -P models/coco-ssd/ https://storage.googleapis.com/tfjs-models/savedmodel/ssdlite_mobilenet_v2/group1-shard3of5
wget -q --show-progress -P models/coco-ssd/ https://storage.googleapis.com/tfjs-models/savedmodel/ssdlite_mobilenet_v2/group1-shard4of5
wget -q --show-progress -P models/coco-ssd/ https://storage.googleapis.com/tfjs-models/savedmodel/ssdlite_mobilenet_v2/group1-shard5of5

echo "Get MobileNetv1 model"
rm -f models/mobilenetv1/*
wget -q --show-progress -P models/mobilenetv1/ https://storage.googleapis.com/tfhub-tfjs-modules/google/imagenet/mobilenet_v1_100_224/classification/1/model.json
wget -q --show-progress -P models/mobilenetv1/ https://storage.googleapis.com/tfhub-tfjs-modules/google/imagenet/mobilenet_v1_100_224/classification/1/group1-shard1of5 
wget -q --show-progress -P models/mobilenetv1/ https://storage.googleapis.com/tfhub-tfjs-modules/google/imagenet/mobilenet_v1_100_224/classification/1/group1-shard2of5 
wget -q --show-progress -P models/mobilenetv1/ https://storage.googleapis.com/tfhub-tfjs-modules/google/imagenet/mobilenet_v1_100_224/classification/1/group1-shard3of5 
wget -q --show-progress -P models/mobilenetv1/ https://storage.googleapis.com/tfhub-tfjs-modules/google/imagenet/mobilenet_v1_100_224/classification/1/group1-shard4of5 
wget -q --show-progress -P models/mobilenetv1/ https://storage.googleapis.com/tfhub-tfjs-modules/google/imagenet/mobilenet_v1_100_224/classification/1/group1-shard5of5 

# echo "Get MobileNetv2 model"
# rm -f models/mobilenetv2/*
# wget -q --show-progress -P models/mobilenetv2/ https://storage.googleapis.com/tfhub-tfjs-modules/google/imagenet/mobilenet_v2_100_224/classification/1/model.json
# wget -q --show-progress -P models/mobilenetv2/ https://storage.googleapis.com/tfhub-tfjs-modules/google/imagenet/mobilenet_v2_100_224/classification/1/group1-shard1of4 
# wget -q --show-progress -P models/mobilenetv2/ https://storage.googleapis.com/tfhub-tfjs-modules/google/imagenet/mobilenet_v2_100_224/classification/1/group1-shard2of4 
# wget -q --show-progress -P models/mobilenetv2/ https://storage.googleapis.com/tfhub-tfjs-modules/google/imagenet/mobilenet_v2_100_224/classification/1/group1-shard3of4 
# wget -q --show-progress -P models/mobilenetv2/ https://storage.googleapis.com/tfhub-tfjs-modules/google/imagenet/mobilenet_v2_100_224/classification/1/group1-shard4of4 

tfjs-nodes.js
module.exports = function (RED) {
  // Common stuff for all nodes
  const tf = require('@tensorflow/tfjs-node')
  const fs = require('fs')
  const path = require('path')
  const express = require('express')
  const compression = require('compression')
  global.fetch = require('node-fetch')

  RED.httpNode.use(compression())
  RED.httpNode.use('/coco', express.static(__dirname + '/models/coco-ssd'))
  RED.httpNode.use('/mobilenet', express.static(__dirname + '/models/mobilenetv1'))

  function setNodeStatus (node, status) {
    switch (status) {
      case 'modelReady':
        node.status({ fill: 'green', shape: 'dot', text: 'ready' })
        break
      case 'modelLoading':
        node.status({ fill: 'yellow', shape: 'ring', text: 'loading model...' })
        break
      case 'infering':
        node.status({ fill: 'blue', shape: 'ring', text: 'infering...' })
        break
      case 'modelError':
        node.status({ fill: 'red', shape: 'dot', text: 'model error' })
        break
      case 'error':
        node.status({ fill: 'red', shape: 'dot', text: 'error' })
        break
      case 'close':
        node.status({})
        break
      default:
        node.status({ fill: 'grey', shape: 'dot', text: status })
    }
  }

  async function inputNodeHandler (node, msg, params) {
    try {
      node.success = false
      if (node.ready) {
        let image = msg.payload
        if (typeof image === 'string') { image = fs.readFileSync(image) } // If it's a string assume it's a filename
        const results = await node.inferImage(image, params)
        setNodeStatus(node, 'modelReady')
        node.success = true
        return results
      } else {
        node.error('model is not ready')
      }
    } catch (error) {
      node.error(error)
      setNodeStatus(node, 'error')
    }
  }

  function filterThreshold (results, threshold) {
    for (let i = 0; i < results.length; i++) {
      if (results[i].score < (threshold / 100)) {
        results.splice(i, 1)
        i--
      }
    }
    return results
  }

  function changeKeyResults (results) {
    const out = []
    for (let i = 0; i < results.length; i++) {
      out.push({
        class: results[i].className,
        score: results[i].probability
      })
    }
    return out
  }

  function countClasses (results) {
    const classes = {}
    for (let i = 0; i < results.length; i++) {
      classes[results[i].class] = (classes[results[i].class] || 0) + 1
    }
    return classes
  }


  // Predict Model
  function tensorflowPredict (config) {
    RED.nodes.createNode(this, config)
    this.mode = config.mode
    this.modelUrl = config.modelUrl
    this.localModel = config.localModel
    this.passthru = config.passthru
    this.params = {}

    var node = this

    async function loadModel () {
      setNodeStatus(node, 'modelLoading')
      try {
        node.ready = false
        if (node.mode === 'online') {
          if (node.modelUrl === '') {
            setNodeStatus(node, 'set a New URL')
            return
          } else {
            node.model = await tf.loadLayersModel(node.modelUrl)
            const shape = node.model.inputs[0].shape
            shape.shift()
            // node.log('input model shape: ' + shape)
            shape.unshift(1)
            node.model.predict(tf.zeros(shape)).dispose()
            node.shape = shape
          }
        } else {
          node.model = await tf.loadLayersModel(node.localModel)
        //   setNodeStatus(node, 'mode not supported')
        //   node.ready = false
        //   return
          // var url = path.join('.node-red', 'node_modules', 'node-red-contrib-tfjs-nodes')
          // var modelUrl = 'file://../' + url + '/models/' + node.localModel + '/model.json'
          // node.model = await mobilenet.load({ modelUrl: modelUrl })
        }
        node.ready = true
        setNodeStatus(node, 'modelReady')
      } catch (error) {
        setNodeStatus(node, 'modelError')
        node.error(error)
      }
    }

    node.inferImage = async function (image, params) {
      setNodeStatus(node, 'infering')
      let tensorImage = tf.node.decodeImage(image, node.shape[3])
      // Rescale the image to fit the wanted shape
      const scaledTensorImage = tf.image.resizeBilinear(tensorImage, [node.shape[1], node.shape[2]], true)
      const offset = tf.scalar(127.5)
      // Normalize the image from [0, 255] to [-1, 1].
      const normalized = scaledTensorImage.sub(offset).div(offset)
      tensorImage = normalized.reshape(node.shape)

      const tensorResult = node.model.predict(tensorImage)
      const indexMax = tensorResult.argMax(1).dataSync()[0]
      const resultsArray = Array.from(tensorResult.dataSync())

      tf.dispose(tensorImage) // Free space

      const results = {
        resultsArray: resultsArray,
        indexMax: indexMax
      }

      return results
    }

    loadModel()

    node.on('input', function (msg) {
      if (node.passthru === true) { msg.image = msg.payload }
      const dynamicParams = { }
      inputNodeHandler(node, msg, dynamicParams).then(
        function (results) {
          if (node.success) {
            msg.indexMax = results.indexMax
            msg.payload = results.resultsArray
            node.send(msg)
          }
        })
    })

    node.on('close', function () { setNodeStatus(node, 'close') })
  }


  // MobileNet Model
  function tensorflowMobilenet (config) {
    const mobilenet = require('@tensorflow-models/mobilenet')

    RED.nodes.createNode(this, config)
    this.mode = config.mode
    this.modelUrl = config.modelUrl
    this.passthru = config.passthru
    this.params = {
      threshold: config.threshold
    }

    var node = this

    async function loadModel () {
      setNodeStatus(node, 'modelLoading')
      try {
        node.ready = false
        if (node.modelUrl === '') {
          node.model = await mobilenet.load()
        } else {
          node.model = await mobilenet.load({ modelUrl: node.modelUrl, version: 1, alpha: 1.0 })
        }
        node.ready = true
        setNodeStatus(node, 'modelReady')
      } catch (error) {
        setNodeStatus(node, 'modelError')
        node.error(error)
      }
    }

    node.inferImage = async function (image, params) {
      setNodeStatus(node, 'infering')
      const tensorImage = tf.node.decodeImage(image)
      const classification = await node.model.classify(tensorImage)
      const filteredClassification = filterThreshold(changeKeyResults(classification), params.threshold)

      tf.dispose(tensorImage) // Free space

      const results = {
        classification: filteredClassification
      }

      return results
    }

    loadModel()

    node.on('input', function (msg) {
      if (node.passthru === true) { msg.image = msg.payload }
      msg.threshold = parseInt(msg.threshold || node.params.threshold) // Adds to msg if it doesn't exist yet
      const dynamicParams = { threshold:msg.threshold }

      inputNodeHandler(node, msg, dynamicParams).then(
        function (results) {
          if (node.success) {
            msg.payload = results.classification
            node.send(msg)
          }
        })
    })

    node.on('close', function () { setNodeStatus(node, 'close') })
  }


  // CoCo SSD Model
  function tensorflowCocoSsd (config) {
    const cocoSsd = require('@tensorflow-models/coco-ssd')

    RED.nodes.createNode(this, config)
    this.modelUrl = config.modelUrl
    this.passthru = config.passthru
    this.params = {
      threshold: config.threshold,
      maxDetections: config.maxDetections
    }

    var node = this

    async function loadModel () {
      setNodeStatus(node, 'modelLoading')
      try {
        node.ready = false
        if (node.modelUrl === '') {
          node.model = await cocoSsd.load()
        } else {
          node.model = await cocoSsd.load({ modelUrl:node.modelUrl })
        }
        node.ready = true
        setNodeStatus(node, 'modelReady')
      } catch (error) {
        setNodeStatus(node, 'modelError')
        node.error(error)
      }
    }

    node.inferImage = async function (image, params) {
      setNodeStatus(node, 'infering')
      const tensorImage = tf.node.decodeImage(image)
      const detections = await node.model.detect(tensorImage, params.maxDetections)
      const filteredDetections = filterThreshold(detections, params.threshold) // Deep copy
      const classes = countClasses(filteredDetections)
      // const filteredResultsModified =

      tf.dispose(tensorImage) // Free space

      const results = {
        filteredDetections: filteredDetections,
        classes: classes
      }

      return results
    }

    loadModel()

    node.on('input', function (msg) {
      if (node.passthru === true) { msg.image = msg.payload }
      msg.threshold = parseInt(msg.threshold || node.params.threshold) // Adds to msg if it doesn't exist yet
      msg.maxDetections = parseInt(msg.maxDetections || node.params.maxDetections) // Adds to msg if it doesn't exist yet

      const dynamicParams = { threshold:msg.threshold, maxDetections:msg.maxDetections }

      inputNodeHandler(node, msg, dynamicParams).then(
        function (results) {
          if (node.success) {
            msg.payload = results.filteredDetections
            msg.classes = results.classes
            node.send(msg)
          }
        }
      )
    })

    node.on('close', function () { setNodeStatus(node, 'close') })
  }


  // PoseNet Model
  function tensorflowPosenet (config) {
    const posenet = require('@tensorflow-models/posenet')

    RED.nodes.createNode(this, config)
    this.modelUrl = config.modelUrl
    this.passthru = config.passthru
    this.params = {
      threshold: config.threshold,
      maxDetections: config.maxDetections
    }

    var node = this

    async function loadModel () {
      setNodeStatus(node, 'modelLoading')
      try {
        node.ready = false
        if (node.modelUrl === '') {
          node.model = await posenet.load()
        } else {
          node.model = await posenet.load({ modelUrl: node.modelUrl })
        }
        node.ready = true
        setNodeStatus(node, 'modelReady')
      } catch (error) {
        setNodeStatus(node, 'modelError')
        node.error(error)
      }
    }

    node.inferImage = async function (image, params) {
      setNodeStatus(node, 'infering')
      const tensorImage = tf.node.decodeImage(image)
      const poses = await node.model.estimateMultiplePoses(tensorImage, {
        flipHorizontal: false,
        maxDetections: params.maxDetections,
        scoreThreshold: params.threshold / 100,
        nmsRadius: 20
      })

      const filteredResults = filterThreshold(poses, params.threshold)

      tf.dispose(tensorImage) // Free space

      const results = {
        filteredResults: filteredResults,
        classes: filteredResults.length ? { person: filteredResults.length } : {}
      }
      return results
    }

    loadModel()

    node.on('input', function (msg) {
      if (node.passthru === true) { msg.image = msg.payload }
      msg.threshold = parseInt(msg.threshold || node.params.threshold) // Adds to msg if it doesn't exist yet
      msg.maxDetections = parseInt(msg.maxDetections || node.params.maxDetections) // Adds to msg if it doesn't exist yet

      const dynamicParams = { threshold:msg.threshold, maxDetections:msg.maxDetections }

      inputNodeHandler(node, msg, dynamicParams).then(
        function (results) {
          if (node.success) {
            msg.payload = results.filteredResults
            msg.classes = results.classes
            node.send(msg)
          }
        }
      )
    })

    node.on('close', function () { setNodeStatus(node, 'close') })
  }

  // Register Node-RED nodes
  RED.nodes.registerType('tensorflowPredict', tensorflowPredict)
  RED.nodes.registerType('tensorflowMobilenet', tensorflowMobilenet)
  RED.nodes.registerType('tensorflowCocoSsd', tensorflowCocoSsd)
  RED.nodes.registerType('tensorflowPosenet', tensorflowPosenet)
}
